src/hdskins/java/com/voxelmodpack/hdskins/resources/SkinResourceManager.java

```diff
@@ -99,31 +97,33 @@ public ResourceLocation getConvertedResource(@Nullable ResourceLocation res) {
         return converted.get(res);
     }
 
-    private void loadSkinResource(@Nullable final ResourceLocation res) {
+    /**
+     * read and convert in a new thread
+     */
+    private void loadSkinResource(@Nullable ResourceLocation res) {
         if (res != null) {
-            // read and convert in a new thread
-            this.inProgress.computeIfAbsent(res, r -> CompletableFuture.supplyAsync(new ImageLoader(r), executor)
-                    .whenComplete((loc, t) -> {
-                        if (loc != null)
-                            converted.put(res, loc);
-                        else {
-                            LogManager.getLogger().warn("Errored while processing {}. Using original.", res, t);
-                            converted.put(res, res);
-                        }
-                    }));
-
-
+            inProgress.computeIfAbsent(res, r -> scheduleConvertion(res, r));
         }
+    }
 
+    private Future<ResourceLocation> scheduleConvertion(ResourceLocation res, ResourceLocation r) {
```

Comment:
do you realize that `r` and `res` are the same object? If you remove one parameter, you can make this just `computeIfAbsent(this::scheduleConvertion)`

Reply:
They were treated as different objects in the original.

Reply:
if you look at the source of `computeIfAbsent`, you'll see this is true.

Reply:
Oh, I think there's something not been interpreted correctly here. The body of the method used a `res` and a `loc` parameter.

Reply:
`loc` is what you renamed to `result`.

Reply:
Yeah, `loc` and `t` became the infinitely more descriptive `result` and `error`. So I guess you're right?

Reply:
Why not just do it normally, though?
```            if (!inProgress.containsKey(res)) {
                inProgress.put(res, scheduleConvertion(res, res));
            }```
That's _way_ easier to understand _and_ it cuts out the hidden cost of compile-time generating an inner anonymous class to run-time instantiate a hidden object for a simple check.

Reply:
I always thought of them as extra sugar on top of these:

```
new Thread(new Runnable() {
   @Override
   public void run() {
       ???
       profit();
   }
});
```